### DISCUSSION TITLE
[Newby Question] Restrict outgoing mail to specific domain

### DISCUSSION THREAD

**Discussion by jdtron** [docker-mailserver/docker-mailserver#3529] [Q&A] [answered]
This might possibly be an easy task, but as the title already says, this is my first time ever hosting an email server.

I have a local mail server setup with the docker-mailserver container, that only serves the purpose of sending and accepting mail for my home lab. Everything works like a charm.

It's not exposed to the internet and it's meant to just work with my internal domain, so I wanted to block outgoing mails to other than my internal domain, to avoid bouncing mails to public servers, that reject them anyways.

I already searched for a solution and stumbled upon possibilities using transport maps and the `smtp_recipient_restrictions` setting.
But none of them worked.

Is there a (possibly simple) solution, so that my server rejects mail to other domains than my internal one?
Thanks in advance!

**Comment by polarathene** [comment, marked as answer]
## Solution

> transport maps

This [seems like the way to go](https://serverfault.com/questions/94168/postfix-on-development-server-allow-mail-to-be-sent-to-only-one-domain/774940#774940), although there is no official DMS support for making that easier, it's not too difficult for you to configure with us 😎 

### Custom config files (_without runtime scripts_)

Use our [`postfix-main.cf` override](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/advanced/override-defaults/postfix/) to enable [`transport_maps`](https://www.postfix.org/postconf.5.html#transport_maps):

```
# text is used instead of hash, this should work and not require the postmap + reload step:
transport_maps = text:/etc/postfix/transport
```

Then volume mount the config to `/etc/postfix/transport`, or alternatively, update that path above to point to your config volume (`/tmp/docker-mailserver/path/to/transport`):

```
.example.com   :
example.com    :
*              discard:
```

[Postfix docs for `transport` table](http://www.postfix.org/transport.5.html) explain most of it.
- The `*` is ok (see "Table Search Order")
- The `:` for the first two lines is detailed in "Result Format" (_5th paragraph, null transport + null nexthop == "do not change"_)
- The above snippet is very similar to the first "Example" snippet.
- [`discard`](https://www.postfix.org/discard.8.html) is a different transport to direct all other mail to (_configured in `/etc/postfix/master.cf`_). It will do as it implies and discard the mail instead of delivering it.

### Via `user-patches.sh` (_all in one file_)

If the above doesn't work, and you need to do the `hash:` table type with the [`postmap`](http://www.postfix.org/postmap.1.html) command, you can handle this with our [`user-patches.sh`](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/advanced/override-defaults/user-patches/), that'll run after startup of DMS is finished, but just before we start Postfix (so no reload necessary).

**`user-patches.sh`:**

```bash
#! /bin/bash

# Update `main.cf` directly instead of `postfix-main.cf` override:
postconf 'transport_maps = hash:/etc/postfix/transport'

# NOTE: If you prefer to copy separate file instead of using below heredoc, use:
# cp /tmp/docker-mailserver/transport /etc/postfix/transport

# HereDoc to create the file inline:
cat > /etc/postfix/transport << EOF
.example.com   :
example.com    :
*              discard:
EOF

# hash table type requires running `postmap` command to update:
postmap /etc/postfix/transport
```

---

## Alternatives

> `smtp_recipient_restrictions`

That is not a valid setting, there's only `smtpd_recipient_restrictions`, but that will only limit inbound mail (_port 25, or submission(s) ports 587 and 465_), not mail Postfix could send out itself or permit through other means (_local mail submission such as `sendmail` command I think goes straight to the mail queue bypassing any restriction?_).

The order of the restrictions is also important, and probably a bit annoying to tamper with when DMS modifies it in a few places during startup.

[This answer](https://serverfault.com/questions/412638/block-outgoing-mail-to-specific-address-using-postfix/832772#832772) also notes that it doesn't seem like `smtpd_recipient_restrictions` would be as viable as `transport_maps`.

> Is there a (possibly simple) solution, so that my server rejects mail to other domains than my internal one?

You could use your own local DNS, such as with a CoreDNS container and really limit it for internal use, but that's probably not as easy 😅

**Reply by jdtron** [reply]
Hi @polarathene and thanks vor the very detailed explaination!

I tried your suggested solutions in order you wrote them down.

**Config files**
For some reason Postfix doesn't seem to like the `text` map type. I got an error in my log, saying that it's not supported.

**user-patches.sh**
This one worked immediately. I chose to have the transport config in a separate file and let the script copy it over, as you suggested. It works like a charm.

Thank you very much for helping me understand better how stuff works and making me able to get things to work! 🥳

**Reply by polarathene** [reply]
> I got an error in my log, saying that it's not supported.

Yeah I figured it didn't support that for this setting based on the Postfix docs and linked ServerFault answer. Sometimes it is which is a little convenient. Thanks for letting me know 👍 

---

> This one worked immediately.
> I chose to have the transport config in a separate file and let the script copy it over, as you suggested. It works like a charm.

Fantastic! 🥳 

There have been similar discussions in the past for using `transport_maps` for a `noreply@example.com` which points to a `discard` (_although the `transport` docs examples show a potentially better way with `error` to send a bounce message back to anyone who does try_). Maybe DMS will get some integration some day :)

---

> Thank you very much

No problem! Enjoy DMS 😁

**Reply by jdtron** [reply]
>the transport docs examples show a potentially better way with error

This might be a even better approach.
I updated my transport map to use `error` instead of `discard`, so the user is informed, that the mail was not delivered.

>No problem! Enjoy DMS 😁

I do, thanks to people like you and the nice community that built such a greate project!
It's a nice way for people like me, who were never in touch with a mail server, to easily get a running system and dive into the details in a much more enjoyable way.

Thanks to everyone involved for making this project possible and being so friendly and helpful 🙂
